Re-open the log file by filesystem notify

marcoandredinis opened the pull request:
This PR implements a handler for logfile re-open by subscription on the filesystem notifications, once file renamed (which usually happen with logrotate default configuration), log file must be reopened. Added configuration flag to enable or disable the watcher, might be useful if graceful restart should be used instead.

changelog: Re-open log file by filesystem notify rename/remove (useful for logrotate).

Related:
https://github.com/gravitational/teleport/pull/6589
https://github.com/gravitational/teleport/pull/9468
https://github.com/gravitational/customer-sensitive-requests/issues/155

Co-authored-by: @vapopov

marcoandredinis committed:
Re-open the log file on OS Signal trigger

This PR implements an handler for the CONT signal.
The handler will close and re-open the log file.
If the log destination is not a file, this is a no-op.

Useful for logrotate postrotate hook.

vapopov commented:
@marcoandredinis I'm going to change the logic a bit, to test it within SIGHUP signal instead, it might be useful in case we have some active sessions that produce the logs and in the case with graceful restart, we should guarantee that the logs from still active sessions not going to split into two log files, so process keep writing in the same file until we can restart

vapopov commented:
@marcoandredinis I've checked the implementation with SIGHUP, SIGCONT you cover both cases so everything looks good

espadolini commented:
Does that imply that any logrotate hook has to send SIGCONT to all the teleport processes that might've spawned by HUP?

Is that true of all the exec and port forward processes too?

r0mant reviewed (commented):
@vapopov @marcoandredinis @espadolini I just had a thought regarding this - instead of hogging another signal here (which are a scarce resource for us), what do you think about adding a new "post logrotate" command to the new [debug service](https://goteleport.com/docs/management/diagnostics/profiles/#using-the-debug-service) that @gabrielcorado recently implemented?

It is enabled by default and already has the logic for changing log level for a running process so having ability to reopen log files seems like a natural addition. We could add another[ API handler](https://github.com/gravitational/teleport/blob/6c1be32b9a977acd315e34f6098761da5663d58c/lib/srv/debug/handlers.go#L42-L54) to it and a corresponding [CLI command](https://github.com/gravitational/teleport/blob/master/tool/teleport/common/debug.go), say `teleport debug post-logrotate` that would be set as a post-rotate hook for logrotate.

WDYGT?

espadolini commented:
The debug service doesn't have a way to make _all_ running Teleport processes reopen their log file, which is possible instead with a `killall teleport` (which isn't particularly great either, admittedly, but anyone relying on logrotate in this day and age is probably fine with the racy behavior).

r0mant commented:
> The debug service doesn't have a way to make all running Teleport processes reopen their log file, which is possible instead with a killall teleport (which isn't particularly great either, admittedly, but anyone relying on logrotate in this day and age is probably fine with the racy behavior).

That's fair. I'm slightly concerned about us just picking random signals for purposes they're not originally intended for meaning e.g. CONT being supposed to resume a suspended process (not saying we have or will have a use-case for it in teleport but still feels a bit off) but I don't see any other signals that would be more appropriate. Interestingly enough, it appears you can actually send payload with the signal using `-q` flag (which sends it using `sigqueue` rather than `kill` according to [man page](https://man7.org/linux/man-pages/man1/kill.1.html)) but Go [doesn't support it](https://github.com/golang/go/issues/9764).

marcoandredinis commented on lib/config/configuration.go:
```diff
@@ -779,14 +780,22 @@ func applyLogConfig(loggerConfig Log, cfg *servicecfg.Config) error {
 	default:
 		// assume it's a file path:
-		logFile, err := os.Create(loggerConfig.Output)
+		var flag int = os.O_WRONLY | os.O_CREATE | os.O_APPEND
+		var mode = fs.FileMode(0666)
```
We should move this to consts

espadolini commented:
> I don't see any other signals that would be more appropriate
It's either CONT or WINCH; the latter can spuriously trigger if Teleport is executed in a terminal, but if we're doing signal hackery that's better than manually sending CONT?

Alternatively, we could run the debug service on the standard socket as well as a per-execution socket (random name, `pid.processid`, something like that) and that'd actually have the benefit of allowing log level changes and pprof to be used on terminating processes. The `teleport debug reopen-logfile` command would just have to loop over all the execution-specific debug sockets rather than just the main one (and there'd be a point to also doing that for the set loglevel command). That's likely way more reliable than firing some random signal at all processes that happened to match `teleport`. WDYT?

r0mant commented:
That might work but I just had (yet) another thought. Can we avoid having to listen for any new signals or introducing new commands altogether by detecting that the log file has been rotated and reopening it automatically?

For example, can we use [fsnotify](https://pkg.go.dev/github.com/fsnotify/fsnotify) in combination with logrorate's [create](https://man.archlinux.org/man/logrotate.8.en#create) and/or [renamecopy](https://man.archlinux.org/man/logrotate.8.en#renamecopy) options to detect that the log file was renamed or a new one was created and reopen it then?

espadolini commented:
If it works in most situations I'd be into that.

It might not work if the log file directory is mounted over NFS and the likes, but I don't think we care that much about setups like that.

r0mant commented:
@vapopov Can you give it a shot and see if it's a viable approach?

vapopov commented:
@r0mant I will prepare prototype with `fsnotify` (we should probably have a config option for this one, in case the user wants to have a choice with `SIGHUP` not to react on move file by logrotate)

vapopov committed:
Merge remote-tracking branch 'origin/master' into marco/logrotate

vapopov committed:
signal replaced with the fsnotify

espadolini commented on lib/utils/log/writer.go:
```diff
@@ -79,7 +82,53 @@ func (s *FileSharedWriter) Reopen() (err error) {
+		return trace.Wrap(err)
+	}
+
+	err = watcher.Add(s.Name())
+	if err != nil {
+		return trace.Wrap(err)
+	}
```
I believe that the goroutine that services `watcher.Events` and `watcher.Errors` must be started before any method calls on `watcher`, or you risk a deadlock.

espadolini commented on lib/utils/log/writer.go:
```diff
@@ -43,3 +50,85 @@ func (s *SharedWriter) Write(p []byte) (int, error) {
+// SharedWriter must be used for log destinations that don't have the reopen requirement, like stdout and stderr.
+// This is thread safe.
+type FileSharedWriter struct {
+	*os.File
```
Avoid embedding `*os.File`, it's never safe to do anything with it unless you are the `FileSharedWriter` anyway.

You might get away with not using a lock (which is always sketchy, when doing I/O) if you use an `atomic.Pointer[os.File]` here instead.

espadolini commented on lib/utils/log/writer.go:
```diff
@@ -43,3 +50,85 @@ func (s *SharedWriter) Write(p []byte) (int, error) {
+		return trace.Wrap(err)
+	}
+
+	err = watcher.Add(s.Name())
```
Are we guaranteed to get all relevant events by watching the file itself, or should we be watching the directory instead?

vapopov replied:
By default `fsnotify` subscribes to all events (`Write`, `Remove`, `Rename`, `Chmod`), `Create` event should work in scope of directory only. 

Btw if we start watching the file and rename it, we stop receiving events. So might be watching directory be better option

espadolini commented on lib/service/signals.go:
```diff
@@ -76,6 +76,7 @@ var teleportSignals = []os.Signal{
 	syscall.SIGUSR1, // log process diagnostic info
 	syscall.SIGUSR2, // initiate process restart procedure
 	syscall.SIGHUP,  // graceful restart procedure
+	syscall.SIGCONT, // Re-open the log file
```
```suggestion
```

espadolini reviewed (commented):
As written, multiple `TeleportProcess`es configured to watch for log file reopens will all independently watch and reopen their log file, but there's pretty much no sensible point in time where we can stop watching the log file.

Considering that the log file is effectively a global resource (and will be for the forseeable future, we have way too many places where we just refer to the global slog or logrus logger), should the watcher mechanism exist as a package-level singleton instead?

vapopov committed:
code review changes:
- refactored to launch watcher with process context
- make file shared writer as singleton
- add test for logrotate case

vapopov committed:
add locking for write, reopen

vapopov committed:
- configuration for log section update
- test simplification

vapopov committed:
Merge remote-tracking branch 'origin/master' into marco/logrotate

espadolini commented on lib/utils/log/file_writer.go:
```diff
@@ -0,0 +1,145 @@
+					return
+				}
+				slog.ErrorContext(ctx, "Error received on logger watcher", "error", err)
+			case <-ctx.Done():
+				if err := watcher.Close(); err != nil {
+					slog.ErrorContext(context.Background(), "Failed to close file watcher", "error", err)
+				}
```
That's a busyloop until one of events and errors is closed; ability to ever stop watching the file notwithstanding, why not tie the context to the watcher close with a `context.AfterFunc`?

vapopov commented on lib/config/configuration.go:
```diff
@@ -779,14 +787,22 @@ func applyLogConfig(loggerConfig Log, cfg *servicecfg.Config) error {
 			return trace.Wrap(err, "failed to create the log file")
 		}
-		w = logFile
+		w, err = logutils.InitFileSharedWriter(logFile, logFileDefaultFlag, logFileDefaultMode)
```
good point, also was thinking about this one, going to check the case with: 
`enable watcher -> start -> disable -> restart`

vapopov committed:
code review changes:
- add closing for file shared writer
- make runWatcherFunc not exported

vapopov committed:
linter fixes

espadolini commented on lib/utils/log/file_writer.go:
```diff
@@ -0,0 +1,182 @@
+		return trace.Wrap(err)
+	}
+
+	oldWatcher := s.watcher.Swap(watcher)
+	if oldWatcher != nil {
+		return trace.Wrap(oldWatcher.Close())
```
Is the watcher supposed to live until the context, or until it's swapped out? Having both is incredibly confusing to understand what's actually going on.

vapopov replied:
We should have only one watcher per shared writer, since we don't block ability to call `RunWatcherReopen` many times, or close it `FileSharedWriter.Close()` we have to stop previous one

espadolini commented on lib/utils/log/file_writer.go:
```diff
@@ -0,0 +1,182 @@
+// FileSharedWriter is similar to SharedWriter except that it requires a `os.File` instead of a `io.Writer`.
+// This is to allow the File reopen required by logrotate and similar tools.
+type FileSharedWriter struct {
+	logFile  atomic.Pointer[os.File]
+	watcher  atomic.Pointer[fsnotify.Watcher]
```
What's the point of the atomics if every interaction is serialized behind a lock?

vapopov replied:
yes would be better to return lock only, since atomic not cover all cases when we in parallel call `Write`, `Reopen`, this atomic must cover only `runWatcherFunc` so will add lock there

espadolini commented on lib/utils/log/file_writer.go:
```diff
@@ -0,0 +1,182 @@
+				if !ok {
+					return
+				}
+				slog.ErrorContext(ctx, "Error received on logger watcher", "error", err)
```
Do we need to do anything else if we get an error, other than log it?

espadolini replied:
> if we move the log directory somewhere else, we will not able to re-open file since there are no directory, so possible solution it is re-create it

We don't create the directories when we open the log file the first time, right? I'm not sure what the behavior should be in that case, but it seems really awkward to watch for all the directories in the supposed path just to figure out when the file can be reopened, so maybe that's just not our problem and if the directory is moved or deleted we just give up on watching?

vapopov replied:
When we start watching the directory, we actually get event if this specific directory renamed as well, agree with you about creation since we don't know parameter to create it, just was suggestion. If directory is moved `fsnotify` removes this path from watch list

tigrato replied:
Can we pick the initial fsmode of the parent directory and create it using the same value?

vapopov replied:
@tigrato it might unpredictable behavior, since we keep writing in the moved file with directory (reopen fails) we should be good just to raise an error

espadolini commented on lib/utils/log/file_writer.go:
```diff
@@ -0,0 +1,182 @@
+				if !ok {
+					return
+				}
+				if s.logFile.Load().Name() == event.Name && (event.Has(fsnotify.Rename) || event.Has(fsnotify.Remove)) {
```
Should we recreate the file if it's not there when we try to reopen it, or should we only reopen the file once something else puts a new file in its place?

vapopov replied:
we use `os.O_CREATE` flag to re-open file, so we should create it always

espadolini commented on lib/utils/log/file_writer.go:
```diff
@@ -0,0 +1,182 @@
+	initLock.Lock()
+	defer initLock.Unlock()
+
+	if fileSharedWriter != nil {
+		return nil, trace.BadParameter("file shared writer already initialized")
+	}
```
Should a new invocation of `InitFileSharedWriter` just override the previous one instead of erroring out? Or can we rely on `applyLogConfig` only ever getting called once? I think that even during an in-process restart the configuration won't get reparsed, so we might be good here, but it's just a bit weird especially for tests.

vapopov replied:
We always clear the state in applyConfig by `logutils.CloseFileSharedWriter()` then do `logutils.InitFileSharedWriter` only if in configuration we still have output as a file path. I've tested it with the SIGHUP signals, so we re-read the configuration and trigger applyConfig, also I've checked exitContext closing to clean up the state

vapopov commented on lib/config/configuration.go:
```diff
@@ -71,6 +72,13 @@ import (
 
+const (
+	// logFileDefaultMode is the preferred permissions mode for log file.
+	logFileDefaultMode fs.FileMode = 0o666
```
Previously it was `os.Create` with 0666, so I left the same mode
```
func Create(name string) (*File, error) {
	return OpenFile(name, O_RDWR|O_CREATE|O_TRUNC, 0666)
}
```
would be better to have this in config, will change to 0644

vapopov committed:
change default file mode for log file
simplify setting global watcher

vapopov commented:
@espadolini could you please take another look, I simplified package singleton to getter/setter only

Wanted add listener `TeleportExit` event to cancel context of fsnotify watcher, but then realized that in case if we do graceful restart: old teleport process should not disable watcher and also reopen same new file as new process does, otherwise old process keep writing in renamed log file and for instance if logrotate start applying compression/truncate etc., old process must keep appending the old file

vapopov committed:
terminate watcher after override global one
added test for validating global override

espadolini commented on lib/service/service.go:
```diff
@@ -5910,6 +5918,7 @@ func (process *TeleportProcess) StartShutdown(ctx context.Context) context.Conte
 	warnOnErr(process.ExitContext(), process.stopListeners(), process.logger)
 
 	// populate context values
+	fmt.Println(">>>> forks count >>>>>: ", process.forkedTeleportCount.Load())
```
😅 

```suggestion
```

espadolini commented on lib/utils/log/file_writer.go:
```diff
@@ -0,0 +1,170 @@
+	lock sync.Mutex
+}
+
+// SetFileSharedWriter sets global file shared writer.
```
Something like this?

```suggestion
// SetFileSharedWriter sets the global file shared writer, closing the current
// one (if any). The returned error is the error from the closing operation.
```

espadolini commented on lib/utils/log/file_writer.go:
```diff
@@ -0,0 +1,170 @@
+
+// Write writes len(b) bytes from b to the File.
+func (s *FileSharedWriter) Write(b []byte) (int, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	return s.logFile.Load().Write(b)
```
If we acquire the lock in Write (which I guess we have to do anyway, just for the sake of serializing long messages from different logging operations) what's the point of keeping logFile in an atomic pointer? The only place where we interact with logFile without holding the lock is in the watcher loop, and it's only to read the file name, which we could just store in the FileSharedWriter instead, as it's never supposed to change (and, in fact, could probably just be sourced from the file name in the configuration to begin with).

vapopov replied:
agree, I just left it more like preventive for future change, we have concurrent write in `reopen` and `read` in the watcher goroutine, and name never changing for now

vapopov committed:
make log file name static for shared writer

espadolini reviewed (commented):
Sorry to go back on the closing/overwriting again, but since this is a scenario where ownership of the `io.Writer` used as a logger is dubious and fuzzy and getting logs when strange things are occurring is almost precisely the point, we might _actually_ be in the position where using a finalizer is appropriate. What do you think? It gets rid of the global stuff, and all it really needs is a wrapper around `FileSharedWriter` (so that the inner workings don't keep the outer shell alive and the finalizer can actually run) and a finalizer that closes it.

vapopov commented:
@espadolini sure, I will change the logic since there are no context required anymore to shutdown the watcher, the only thing I'm not sure with finalizer that there might be some time gap when finalize function must be called, so there might be situation that we have 2 watcher goroutines (current and previous file shared writer), so previous one might re-open file but never going to be used by write operation and then closed. Current implementation stop watcher more explicit, but in terms of the testing (where we can create many Teleport instances) we always override global one

Btw, in this implementation with global file shared writer, file should be closed by default internal finalizer after replacing global variable, and replacing it in `logger.SetOutput(w)` by new instance, small test: https://go.dev/play/p/A4zZVteRtd-

vapopov committed:
replace logic to use finalizer instead

espadolini commented on lib/utils/log/file_writer.go:
```diff
@@ -0,0 +1,151 @@
+
+// Reopen closes the file and opens it again using APPEND mode.
+func (s *FileSharedWriter) Reopen() (err error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if err := s.file.Close(); err != nil {
+		return trace.Wrap(err)
+	}
+	s.file, err = os.OpenFile(s.logFileName, s.fileFlag, s.fileMode)
```
Open the new file before grabbing the lock and before closing the old one, filesystem operations can block and fail; this should probably also not fail if the old file returns an error on close for whatever reason.

espadolini commented on lib/utils/log/file_writer.go:
```diff
@@ -0,0 +1,151 @@
+}
+
+// Reopen closes the file and opens it again using APPEND mode.
+func (s *FileSharedWriter) Reopen() (err error) {
```
Avoid named return values unless really necessary (due to a defer, typically); if you need an `err` variable just declare it before the first use with `var err error`.

```suggestion
func (s *FileSharedWriter) Reopen() error {
```

espadolini commented on lib/utils/log/file_writer.go:
```diff
@@ -0,0 +1,151 @@
+func (s *FileSharedWriter) Close() error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
```
You should check and set a dedicated "closed" flag here and in `Reopen`, or you're liable to race the file close with the reopen.

espadolini commented on lib/utils/log/writer_finalizer.go:
```diff
@@ -0,0 +1,49 @@
+)
+
+// WriterFinalizer is a wrapper for the [io.WriteCloser] to automate resource cleanup.
+type WriterFinalizer struct {
+	writer io.WriteCloser
+}
```
We only use this with concrete types, we can avoid the virtualization cost I think?

```suggestion
type WriterFinalizer[T io.WriteCloser] struct {
	writer T
}
```

espadolini commented on lib/utils/log/file_writer.go:
```diff
@@ -0,0 +1,151 @@
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if err := s.watcher.Close(); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := s.file.Close(); err != nil {
+		return trace.Wrap(err)
+	}
```
This will leave the file open if the watcher fails to close (and presumably the watcher will forever fail to close cleanly again afterwards, so the file will be open forever). Something like `return trace.NewAggregate(s.watcher.Close(), s.file.Close())` would be better.

vapopov committed:
fix possible locking write while reopen is triggered

espadolini commented on lib/utils/log/file_writer.go:
```diff
@@ -0,0 +1,162 @@
+
+	s.lock.Lock()
+	if s.closed {
+		return trace.NewAggregate(ErrFileSharedWriterClosed, file.Close())
```
I would just
```suggestion
		_ = file.Close()
		return trace.Wrap(ErrFileSharedWriterClosed)
```

espadolini commented on lib/utils/log/file_writer.go:
```diff
@@ -0,0 +1,162 @@
+	}
+
+	s.lock.Lock()
+	if s.closed {
```
```suggestion
	if s.closed {
		s.lock.Unlock()
```

espadolini commented on lib/utils/log/file_writer.go:
```diff
@@ -0,0 +1,162 @@
+	defer s.lock.Unlock()
+
+	if s.closed {
+		return ErrFileSharedWriterClosed
```
```suggestion
		return trace.Wrap(ErrFileSharedWriterClosed)
```

espadolini commented on lib/utils/log/file_writer.go:
```diff
@@ -0,0 +1,162 @@
+
+// Write writes len(b) bytes from b to the File.
+func (s *FileSharedWriter) Write(b []byte) (int, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
```
Should we have a separate lock for the write synchronization and for the rest of the FileSharedWriter machinery?

vapopov replied:
Not sure if we really need it, because we can't write when in goroutine reopen is triggered, same for closing the file. Only `RunWatcherReopen` kind of independent of the writing process, but we call it before file shared writer is assigned to the logger (it might be even moved to constructor)

vapopov committed:
wrap internal error, fix incorrect unlock

r0mant commented on docs/pages/reference/helm-reference/teleport-cluster.mdx:
````diff
@@ -1527,6 +1527,23 @@ See the [Teleport config file reference](../../reference/config.mdx) for more de
     extraFields: ["timestamp", "level"]
   ```
 
+### `log.watch_log_file`
````
As discussed in our 1x1, let's remove this extra knob for now.

r0mant commented on lib/utils/log/file_writer.go:
```diff
@@ -0,0 +1,162 @@
+	}
+	logFile, err := os.OpenFile(logFileName, flag, mode)
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to create the log file")
```
```suggestion
		return nil, trace.ConvertSystemError(err, "failed to create the log file")
```

r0mant commented on lib/utils/log/file_writer.go:
```diff
@@ -0,0 +1,162 @@
+				if s.logFileName == event.Name && (event.Has(fsnotify.Rename) || event.Has(fsnotify.Remove)) {
+					slog.DebugContext(context.Background(), "Log file was moved/removed", "file", event.Name)
+					if err := action(); err != nil {
+						slog.ErrorContext(context.Background(), "Failed to take action", "error", err, "file", event.Name)
```
```suggestion
						slog.ErrorContext(context.Background(), "Failed to reopen file", "error", err, "file", event.Name)
```

r0mant commented on lib/utils/log/file_writer.go:
```diff
@@ -0,0 +1,162 @@
+				if !ok {
+					return
+				}
+				slog.ErrorContext(context.Background(), "Error received on logger watcher", "error", err)
```
Nit: can we pass the context to all slog functions from the caller?

vapopov replied:
unfortunately in configuration process (where we init FileSharedWriter) we don't have any context, I wanted to use  `slog.Error`, but it blacklisted in linter test

r0mant replied:
I just meant you can pass it here from RunWatcherReopen, and RunWatcherReopen can in turn accept the context itself from its caller in applyLogConfig - there you can pass context.Background(), but then if we need to plumb a proper context here later if would be easier to do.

vapopov replied:
got you, it was initially like that, then I removed within watcher closing by context cancelation

@r0mant I've made watcher always enabled, and removed this config flag

r0mant reviewed (commented):
lgtm once you remove the extra knob and with a couple minor suggestions

vapopov committed:
drop configuration flag, make watcher always enabled

vapopov committed:
tune context passing

r0mant reviewed (approved):
lgtm. @vapopov I would only backport to v16, just to be safe.
